### PR TITLE
feat(regional): KSA E-Invoicing and VAT Report

### PR DESCRIPTION
**Pull request by ahmadpak**
This PR manages to generate report for the VAT on Sales and VAT on Purchases for Saudi Arab

Workspace(Accounting) Updated
New Doctype added under Taxes: KSA VAT Setting
New Report added under Report: KSA VAT
QR Code generation on creating new sales invoice

Details: https://github.com/ahmadpak/ksa_vat
Docs: https://docs.erpnext.com/docs/v13/user/manual/en/simplified_ksa_vat_management_and_reporting

### CODE REVIEW

File: erpnext/regional/__init__.py [hunk 1]
```diff
@@ -31,3 +34,72 @@ def create_transaction_log(doc, method):
...
+			_file.save()
+
+			# assigning to document
+			doc.db_set('qr_code', _file.file_url)
+			doc.notify_update()
+
+			break
+
+		else:
+			pass
```
Comment (nextchamp-saqib) [inline erpnext/regional/__init__.py]:
```suggestion
```

File: erpnext/regional/report/ksa_vat/ksa_vat.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+	]
+
+def get_data(filters):
+	data = []
+
+	# Validate if vat settings exist
+	company = filters.get('company')
+	if frappe.db.exists('KSA VAT Setting', company) is None:
+		url = get_url_to_list('KSA VAT Setting')
+		frappe.msgprint(f'Create <a href="{url}">KSA VAT Setting</a> for this company')
```
Comment (nextchamp-saqib) [inline erpnext/regional/report/ksa_vat/ksa_vat.py]:
```suggestion
		frappe.msgprint(_('Create <a href="{}">KSA VAT Setting</a> for this company').format(url))
```

File: erpnext/regional/saudi_arabia/setup.py [hunk 3]
```diff
@@ -2,10 +2,36 @@
...
+from erpnext.regional.saudi_arabia.wizard.operations.setup_ksa_vat_setting import create_ksa_vat_setting
+from frappe.custom.doctype.custom_field.custom_field import create_custom_field
 
 def setup(company=None, patch=True):
-	make_custom_fields()
+	uae_custom_fields()
 	add_print_formats()
+	add_permissions()
+	create_ksa_vat_setting(company)
+	make_custom_fields()
```
Comment (nextchamp-saqib) [inline erpnext/regional/saudi_arabia/setup.py]:
```suggestion
	make_qrcode_field()
```

File: erpnext/regional/saudi_arabia/setup.py [hunk 4]
```diff
@@ -2,10 +2,36 @@
...
+	add_permission('KSA VAT Setting', 'All', 0)
+	for role in ('Accounts Manager', 'Accounts User', 'System Manager'):
+		add_permission('KSA VAT Setting', role, 0)
+		update_permission_property('KSA VAT Setting', role, 0, 'write', 1)
+		update_permission_property('KSA VAT Setting', role, 0, 'create', 1)
+
+	"""Enable KSA VAT Report"""
+	frappe.db.set_value('Report', 'KSA VAT', 'disabled', 0)
+
+def make_custom_fields():
```
Comment (nextchamp-saqib) [inline erpnext/regional/saudi_arabia/setup.py]:
```suggestion
def make_qrcode_field():
```

File: erpnext/hooks.py [hunk 5]
```diff
@@ -262,6 +262,7 @@
 		"validate": "erpnext.regional.india.utils.validate_tax_category"
 	},
 	"Sales Invoice": {
+		"after_insert": "erpnext.regional.create_qr_code",
```
Comment (nextchamp-saqib) [inline erpnext/hooks.py]: `erpnext.regional.create_qr_code` can be moved to `erpnext.regional.saudi_arabia.utils.create_qr_code`?

As we have namespaced all the Indian helper functions to `regional.india`.

Similarly, I think all the new utils should be a part of `saudi_arabia.utils`

Comment (ahmadpak) [reply erpnext/hooks.py]: I saw similar code in __init__.py for `["Nepal", "France"]`  and for `["France", "Germany"]`
Anyways moved it as suggested

File: erpnext/regional/saudi_arabia/utils.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+	# Don't create QR Code if it already exists
+	qr_code = doc.get("qr_code")
+	if qr_code and frappe.db.exists({"doctype": "File", "file_url": qr_code}):
+		return
+
+	fields = frappe.get_meta('Sales Invoice').fields
+	
+	for field in fields:
+		if field.fieldname == 'qr_code' and field.fieldtype == 'Attach Image':
```
Comment (nextchamp-saqib) [inline erpnext/regional/saudi_arabia/utils.py]:
```suggestion
	meta = frappe.get_meta('Sales Invoice')
	for field in meta.get_image_fields():
		if field.fieldname == 'qr_code':
```

File: erpnext/regional/saudi_arabia/utils.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			url = qr_create(url, error='L')
+			url.png(qr_image, scale=2, quiet_zone=1)
+			
+			# making file
+			filename = f"QR-CODE-{doc.name}.png".replace(os.path.sep, "__")
+			_file = frappe.get_doc({
+				"doctype": "File",
+				"file_name": filename,
+				"is_private": 0,
+				"content": qr_image.getvalue()
```
Comment (nextchamp-saqib) [inline erpnext/regional/saudi_arabia/utils.py]:
```suggestion
				"content": qr_image.getvalue(),
				"attached_to_doctype": doc.get("doctype"),
				"attached_to_name": doc.get("name"),
				"attached_to_field": "qr_code"
```

File: erpnext/regional/saudi_arabia/utils.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	"""Delete QR Code on deleted sales invoice"""
+	
+	region = get_region(doc.company)
+	if region not in ['Saudi Arabia']:
+		return
+
+	if hasattr(doc, 'qr_code'):
+		if doc.get('qr_code'):
+			file_doc = frappe.get_list('File', {
+				'file_url': doc.qr_code
```
Comment (nextchamp-saqib) [inline erpnext/regional/saudi_arabia/utils.py]:
```suggestion
				'file_url': doc.get('qr_code')
```

File: erpnext/regional/doctype/ksa_vat_setting/ksa_vat_setting_list.js [hunk 9]
```diff
@@ -0,0 +1,5 @@
+frappe.listview_settings['KSA VAT Setting'] = {
+	onload () {
+		frappe.breadcrumbs.add('Accounts');
+	}
+}
```
Comment (nextchamp-saqib) [inline erpnext/regional/doctype/ksa_vat_setting/ksa_vat_setting_list.js:5]:
```suggestion
};
```

**Comment by nextchamp-saqib** [comment]
@ahmadpak Just need basic tests cases for reports & other critical parts of the feature. 

Also, documentation on the custom app seems to be a little overwhelming, it can be improved later on so that we can move it on docs.erpnext.com and forward our customers to it for the setup.

**Comment by deepeshgarg007** [comment]
@Mergifyio backport version-13-hotfix

**Review by deepeshgarg007 — APPROVED** [review]

**Comment by asamawi** [comment]
When creating a POS invoice a QR code is not generated.
`` in pos invoice format doesn't show anything!
issue created:
https://github.com/ahmadpak/ksa_vat/issues/9

**Comment by sammish** [comment]
@deepeshgarg007 When we can expect this in production bench ?

**Pull request merged**